server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -1582,6 +1589,32 @@ public void clusterStateProcessed(ClusterState oldState, ClusterState newState)
         }
     }
 
+    private IndexMetadata convertLegacyIndex(IndexMetadata snapshotIndexMetadata) {
+        MappingMetadata mappingMetadata = snapshotIndexMetadata.mapping();
+        Map<String, Object> loadedMappingSource = mappingMetadata.rawSourceAsMap();
+
+        // store old mapping under _meta/legacy-mappings
+        Map<String, Object> legacyMapping = new LinkedHashMap<>();
+        boolean sourceOnlySnapshot = snapshotIndexMetadata.getSettings().getAsBoolean("index.source_only", false);
+        if (sourceOnlySnapshot) {
+            // actual mapping is under "_meta" (but strip type first)
+            Object sourceOnlyMeta = mappingMetadata.sourceAsMap().get("_meta");
+            if (sourceOnlyMeta instanceof Map<?, ?> sourceOnlyMetaMap) {
+                legacyMapping.put("legacy-mappings", sourceOnlyMetaMap);
```

Comment:
Should we use `legacy_mappings` instead of `legacy-mappings`?

Reply:
fixed in 939bc51